lesson17.py

```diff
@@ -0,0 +1,139 @@
+
+print("Dog say:", generic_func(dog))
+print("Cat say:", generic_func(cat))
+
+
+# Task 2
+
+
+class Author:
+    def __init__(self, name, country, birthday):
+        self.name = name
+        self.country = country
+        self.birthday = birthday
+        self.books = []
+
+    def __repr__(self):
+        return f"Author ({self.name}, {self.country}, {self.birthday})"
+
+    def __str__(self):
+        return
+
+
+class Book:
+    all_existing_books = 0
+
+    def __init__(self, name, year, author):
+        self.name = name
+        self.year = year
+        self.author = author
+
+    all_existing_books += 1
+
+
+class Library:
+
+    def __init__(self, name):
+        self.name = name
+        self.books = []
+        self.authors = []
+
+    def new_book(self, name: str, year: int, author: Author):
+        all_books = Book(name, year, author)
+        self.books.append(all_books)
+        return all_books
+
+    def group_by_author(self, author: Author):
+        return [book for book in self.books if book.author == author]
+
+    def group_by_year(self, year: int):
+        return [book for book in self.books if book.year == year]
+
+
+# Task 3
+
+
+class Fraction:
+    def __init__(self, numerator: int, denominator: int):
+        if type(numerator) == int and type(denominator) == int:
+            self.numerator = numerator
+            self.denominator = denominator
+        else:
+            raise ValueError('Args must be integer')
+
+    def __str__(self):
+        print(f"{self.numerator}/{self.denominator}")
+        return f"{self.numerator}/{self.denominator}"
+
+    def __add__(self, other):
+        new_numerator = self.numerator * other.denominator + other.numerator * self.denominator
+        new_denominator = self.denominator * other.denominator
+        new_numerator, new_denominator = self.optimized(new_numerator, new_denominator)
+        return Fraction(new_numerator, new_denominator)
+
+    def __eq__(self, other):
+        new_numerator = self.numerator * other.denominator
+        new_denominator = other.numerator * self.denominator
+        return new_numerator == new_denominator
+
+    def __sub__(self, other):
+        new_numerator = self.numerator * other.denominator - other.numerator * self.denominator
+        new_denominator = self.denominator * other.denominator
+        return Fraction(new_numerator, new_denominator)
+
+    def __mul__(self, other):
+        new_numerator = self.numerator * other.numerator
+        new_denominator = self.denominator * other.denominator
+        return Fraction(new_numerator, new_denominator)
+
+    def __truediv__(self, other):
+        if other.numerator == 0:
+            raise ValueError("Ділення на нуль неможливе")
+        new_numerator = self.numerator * other.denominator
+        new_denominator = self.denominator * other.numerator
+        return Fraction(new_numerator, new_denominator)
+
+    @staticmethod
+    def optimized(numerator, denominator):

```

Comment:
It's good, but not good enough. Please try to find greatest common denominator (you can import the `math` module for this purpose